File: integrations/langchain-ts/src/chat_history.ts
/**
 * Xache Chat Message History for LangChain.js
 * Persistent message storage with cryptographic receipts
 */

import { BaseListChatMessageHistory } from '@langchain/core/chat_history';
import {
  BaseMessage,
  HumanMessage,
  AIMessage,
  SystemMessage,
} from '@langchain/core/messages';
import { XacheClient, DID } from '@xache/sdk';
import { randomUUID } from 'crypto';

export interface XacheChatMessageHistoryConfig {
  /** Wallet address for authentication */
  walletAddress: string;
  /** Private key for signing */
  privateKey: string;
  /** Session ID to group messages */
  sessionId?: string;
  /** API URL (defaults to https://api.xache.xyz) */
  apiUrl?: string;
  /** Chain: 'base' or 'solana' */
  chain?: 'base' | 'solana';
  /** Maximum messages to load (default: 1000, set to -1 for unlimited) */
  maxMessages?: number;
  /** Page size for loading messages (default: 100) */
  pageSize?: number;
  /** Request timeout in milliseconds (default: 30000) */
  timeout?: number;
  /** Enable debug logging */
  debug?: boolean;
}

/**
 * Chat message history backed by Xache Protocol.
 *
 * Messages are stored with cryptographic receipts and can persist
 * across sessions.
 *
 * @example
 * ```typescript
 * import { XacheChatMessageHistory } from '@xache/langchain';
 *
 * const history = new XacheChatMessageHistory({
 *   walletAddress: '0x...',
 *   privateKey: '0x...',
 *   sessionId: 'unique-session-id',
 * });
 *
 * await history.addUserMessage('Hello!');
 * await history.addAIMessage('Hi there!');
 *
 * const messages = await history.getMessages();
 * ```
 */
export class XacheChatMessageHistory extends BaseListChatMessageHistory {
  lc_namespace = ['xache', 'chat_history'];

  private client: XacheClient;
  private sessionId: string;
  private messages: BaseMessage[] = [];
  private initialized = false;
  private maxMessages: number;
  private pageSize: number;

  constructor(config: XacheChatMessageHistoryConfig) {
    super();

    const chainPrefix = config.chain === 'solana' ? 'sol' : 'evm';
    const did = `did:agent:${chainPrefix}:${config.walletAddress.toLowerCase()}` as DID;

    this.client = new XacheClient({
      apiUrl: config.apiUrl || 'https://api.xache.xyz',
      did,
      privateKey: config.privateKey,
      timeout: config.timeout,
      debug: config.debug,
    });

    // Use UUID for collision-resistant session IDs
    this.sessionId = config.sessionId || `langchain-${randomUUID()}`;
    this.maxMessages = config.maxMessages ?? 1000;
    this.pageSize = config.pageSize ?? 100;
  }

  /**
   * Get all messages in the history
   */
  async getMessages(): Promise<BaseMessage[]> {
    if (!this.initialized) {
      await this.loadMessages();
    }
    return this.messages;
  }

  /**
   * Add a message to the history
   */
  async addMessage(message: BaseMessage): Promise<void> {
    this.messages.push(message);

    const role = this.getMessageRole(message);
    const content =
      typeof message.content === 'string'
        ? message.content
        : JSON.stringify(message.content);

    await this.client.memory.store({
      data: {
        role,
        content,
        sessionId: this.sessionId,
        timestamp: Date.now(),
      },
      storageTier: 'hot',
      context: `chat:${this.sessionId}`,
      tags: ['chat', 'message', role],
      metadata: {
        sessionId: this.sessionId,
        role,
      },
    });
  }

  /**
   * Add a user message
   */
  async addUserMessage(message: string): Promise<void> {
    await this.addMessage(new HumanMessage(message));
  }

  /**
   * Add an AI message
   */
  async addAIMessage(message: string): Promise<void> {
    await this.addMessage(new AIMessage(message));
  }

  /**
   * Clear all messages from history
   */
  async clear(): Promise<void> {
    this.messages = [];
    // Note: Xache doesn't support deletion - messages remain for audit trail
    // We only clear the local cache
  }

  /**
   * Load messages from Xache storage with pagination support
   */
  private async loadMessages(): Promise<void> {
    try {
      const allMemories: Array<{ storage_key: string; created_at?: string }> = [];
      let offset = 0;
      const effectiveMax = this.maxMessages === -1 ? Infinity : this.maxMessages;

      // Paginate through all messages
      while (allMemories.length < effectiveMax) {
        const result = await this.client.memory.list({
          context: `chat:${this.sessionId}`,
          limit: Math.min(this.pageSize, effectiveMax - allMemories.length),
          offset,
        });

        const memories = Array.isArray(result?.memories) ? result.memories : [];
        if (memories.length === 0) break;

        allMemories.push(...memories);
        offset += memories.length;

        // If we got less than pageSize, we've reached the end
        if (memories.length < this.pageSize) break;
      }

      // Sort by timestamp
      const sortedMemories = [...allMemories].sort((a, b) => {
        const tsA = new Date(a.created_at || 0).getTime();
        const tsB = new Date(b.created_at || 0).getTime();
        return tsA - tsB;
      });

      // For each memory, we need to retrieve the full content
      this.messages = [];
      for (const m of sortedMemories) {
        try {
          const full = await this.client.memory.retrieve({
            storageKey: m.storage_key,
          });
          const data = full.data as { role?: string; content?: string };
          if (data && data.content) {
            switch (data.role) {
              case 'human':
              case 'user':
                this.messages.push(new HumanMessage(data.content));
                break;
              case 'ai':
              case 'assistant':
                this.messages.push(new AIMessage(data.content));
                break;
              case 'system':
                this.messages.push(new SystemMessage(data.content));
                break;
              default:
                this.messages.push(new HumanMessage(data.content));
            }
          }
        } catch (error) {
          // Skip malformed messages but log in debug
          console.debug?.(`[XacheChatMessageHistory] Failed to parse message: ${(error as Error).message}`);
        }
      }

      this.initialized = true;
    } catch (error) {
      // If retrieval fails (e.g., no memories yet), start fresh
      // This is expected for new sessions
      console.debug?.(`[XacheChatMessageHistory] No existing messages found: ${(error as Error).message}`);
      this.messages = [];
      this.initialized = true;
    }
  }

  private getMessageRole(message: BaseMessage): string {
    if (message instanceof HumanMessage) return 'human';
    if (message instanceof AIMessage) return 'ai';
    if (message instanceof SystemMessage) return 'system';
    return 'unknown';
  }
}

File: integrations/langchain-python/xache_langchain/chat_history.py
"""
Xache Chat Message History for LangChain
Persistent chat history with verifiable receipts
"""

import os
from typing import List, Optional
from langchain_core.messages import BaseMessage, HumanMessage, AIMessage, SystemMessage
from langchain_core.chat_history import BaseChatMessageHistory

from xache import XacheClient
from ._async_utils import run_sync


class XacheChatMessageHistory(BaseChatMessageHistory):
    """
    Chat message history backed by Xache Protocol.

    Stores chat messages with cryptographic receipts and optional
    reputation tracking.

    Example:
        ```python
        from xache_langchain import XacheChatMessageHistory
        from langchain.memory import ConversationBufferMemory

        history = XacheChatMessageHistory(
            wallet_address="0x...",
            private_key="0x...",
            session_id="unique-session-id"
        )

        memory = ConversationBufferMemory(chat_memory=history)
        ```
    """

    def __init__(
        self,
        wallet_address: str,
        private_key: str,
        session_id: str,
        api_url: Optional[str] = None,
        chain: str = "base",
    ):
        """
        Initialize Xache chat history.

        Args:
            wallet_address: Wallet address for authentication
            private_key: Private key for signing
            session_id: Unique session identifier
            api_url: Xache API URL
            chain: Blockchain (base, solana)
        """
        self.wallet_address = wallet_address
        self.private_key = private_key
        self.session_id = session_id
        self.api_url = api_url or os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
        self.chain = chain

        # Build DID
        chain_prefix = "sol" if chain == "solana" else "evm"
        self.did = f"did:agent:{chain_prefix}:{wallet_address.lower()}"

        self._client = XacheClient(
            api_url=api_url,
            did=self.did,
            private_key=private_key,
        )

    @property
    def messages(self) -> List[BaseMessage]:
        """Retrieve all messages from Xache"""

        async def _get_messages():
            async with self._client as client:
                # Retrieve memories for this session (filtered by context)
                memories = await client.memory.list(
                    limit=1000,
                    context=f"chat:session:{self.session_id}"
                )

                messages = []
                result_memories = memories.get("memories", []) if isinstance(memories, dict) else memories
                for mem in sorted(result_memories, key=lambda x: x.get("created_at", "")):
                    content = mem.get("content", "")
                    metadata = mem.get("metadata", {})
                    role = metadata.get("role", "human")

                    if role == "human":
                        messages.append(HumanMessage(content=content))
                    elif role == "ai":
                        messages.append(AIMessage(content=content))
                    elif role == "system":
                        messages.append(SystemMessage(content=content))

                return messages

        return run_sync(_get_messages())

    def add_message(self, message: BaseMessage) -> None:
        """Add a message to the history"""

        # Determine role from message type
        if isinstance(message, HumanMessage):
            role = "human"
        elif isinstance(message, AIMessage):
            role = "ai"
        elif isinstance(message, SystemMessage):
            role = "system"
        else:
            role = "unknown"

        async def _add():
            async with self._client as client:
                await client.memory.store(
                    content=message.content,
                    context=f"chat:session:{self.session_id}",
                    metadata={
                        "role": role,
                        "session_id": self.session_id,
                        "source": "langchain",
                        "message_type": message.__class__.__name__,
                    }
                )

        run_sync(_add())

    def add_user_message(self, message: str) -> None:
        """Add a user message"""
        self.add_message(HumanMessage(content=message))

    def add_ai_message(self, message: str) -> None:
        """Add an AI message"""
        self.add_message(AIMessage(content=message))

    def clear(self) -> None:
        """Clear message history (soft delete) using bulk deletion"""
        self.bulk_delete()

    def bulk_delete(self, storage_keys: Optional[List[str]] = None) -> dict:
        """
        Delete multiple messages from history.

        Args:
            storage_keys: List of storage keys to delete. If None, deletes all messages
                         in this session.

        Returns:
            dict with 'deleted' count and 'errors' list
        """
        import asyncio

        async def _bulk_delete():
            async with self._client as client:
                # Get storage keys if not provided
                keys_to_delete = storage_keys
                if keys_to_delete is None:
                    result = await client.memory.list(
                        limit=1000,
                        context=f"chat:session:{self.session_id}"
                    )
                    memories = result.get("memories", []) if isinstance(result, dict) else result
                    keys_to_delete = [
                        mem.get("storage_key") for mem in memories
                        if mem.get("storage_key")
                    ]

                if not keys_to_delete:
                    return {"deleted": 0, "errors": []}

                # Delete in parallel using asyncio.gather with return_exceptions
                async def safe_delete(key):
                    try:
                        await client.memory.delete(key)
                        return {"key": key, "success": True}
                    except Exception as e:
                        return {"key": key, "success": False, "error": str(e)}

                results = await asyncio.gather(
                    *[safe_delete(key) for key in keys_to_delete],
                    return_exceptions=True
                )

                deleted = sum(1 for r in results if isinstance(r, dict) and r.get("success"))
                errors = [
                    r for r in results
                    if isinstance(r, dict) and not r.get("success")
                    or isinstance(r, Exception)
                ]

                return {"deleted": deleted, "errors": errors}

        return run_sync(_bulk_delete())
